### ISSUE TITLE
How to use Redis pub/sub with Quarkus and ReactiveRedisClient?

### ISSUE THREAD

**Issue by digitalsamba** [quarkusio/quarkus#21108]
I am trying to use Redis pub/sub with Quarkus, but I cannot find any documentation about it. I managed to use the Quarkus Redis Extension with Redis Streams xadd and xread, but cannot make Redis pub/sub work. I am using **ReactiveRedisClient** and when I try to subscribe to a redis channel in a Quarkus test I receive following error:

`Error received: io.vertx.core.impl.NoStackTraceThrowable: PubSub command in connection-less mode not allowed`

I found the error is thrown in vertx redis extension **BaseRedisClient** class:
```
@Override
  public Future<@Nullable Response> send(Request command) {
    final Promise<Response> promise = vertx.promise();

    if (command.command().isPubSub()) {
      // mixing pubSub cannot be used on a one-shot operation
      promise.fail("PubSub command in connection-less mode not allowed");
      return promise.future();
    }
  }
```

Generally I find Quarkus would benefit of documenting how to use Redis Pub/Sub cause I found other people are also confused  how to use it exactly.

Thanks in advance for any help!

**Comment by machi1990** [comment]
> `Error received: io.vertx.core.impl.NoStackTraceThrowable: PubSub command in connection-less mode not allowed`

The error means you are trying to use the same client to publish and subscribe which is not allowed by the upstream redis client. /cc @pmlopes

The issue has been reported in the past (I cannot find the reference to it) and our recommendation has been to create two separate clients using the [multiple redis client](https://quarkus.io/guides/redis#multiple-clients-configuration) feature. One create will perform the subscription and the other to publish. 
```
quarkus.redis.hosts=redis://localhost:6379
quarkus.redis.publish.hosts=redis://localhost:6379
``` 

```java

@Inject
ReactiveRedisClient defaultClient; // you can use the default client for subscription


@Inject
@RedisClientName("publish")
ReactiveRedisClient publishingClient; //  you can use this client for publishing
```

**Comment by digitalsamba** [comment]
Thank you for the answer, but I am not sure if this is the case. I literally don't have any code inside my test - it is just a subscribe called and nothing more - no one is publishing or any other action is going on. Could you try such a test by your side with latest Quarkus - I am using Quarkus dev services in test.

```
@QuarkusTest
public class RedisTest {

    @Inject
    ReactiveRedisClient reactiveRedisClient;
    
    @Test
    public void redisTest() {
    	reactiveRedisClient.subscribe(List.of("CHANNEL1", "CHANNEL2"))
	    	.subscribe().with(
					response -> System.out.println("Subscribe Response: " + response),
					error -> System.out.println("Error received: " + error));
    }
}
```

**Comment by digitalsamba** [comment]
I also noticed this inside vertx github - https://github.com/vert-x3/vertx-redis-client/issues/273. I have no idea what is a connection-less mode, but seems subscribe is not allowed in it.

**Comment by machi1990** [comment]
> @machi1990 Thank you for the answer, but I am not sure if this is the case. I literally don't have any code inside my test - it is just a subscribe called and nothing more - no one is publishing or any other action is going on. Could you try such a test by your side with latest Quarkus - I am using Quarkus dev services in test.
> 
> ```
> @QuarkusTest
> public class RedisTest {
> 
>     @Inject
>     ReactiveRedisClient reactiveRedisClient;
>     
>     @Test
>     public void redisTest() {
>     	reactiveRedisClient.subscribe(List.of("CHANNEL1", "CHANNEL2"))
> 	    	.subscribe().with(
> 					response -> System.out.println("Subscribe Response: " + response),
> 					error -> System.out.println("Error received: " + error));
>     }
> }
> ```

Right, I didn't know if you were referencing the subscribe command. I initially thought you were using the pubsub command, my mistake sorry.  

> I also noticed this inside vertx github - [vert-x3/vertx-redis-client#273](https://github.com/vert-x3/vertx-redis-client/issues/273). I have no idea what is a connection-less mode, but seems subscribe is not allowed in it - only publish/pubsub, but not subscribe. If someone could shed on a light on this it would be great.

@pmlopes would you have an idea?

**Comment by DarioFR** [comment]
Hello. I'm currently facing the same issue which prevents even a simple application with a publisher and subscriber from properly working. Any news about a possible solution/workaround?

Thanks.

**Comment by digitalsamba** [comment]
Today I had same time to dig deeper and found the issue. 

Basically the vertx redis client doesn't support pub/sub in connection pooled mode. 

https://vertx.io/docs/vertx-redis-client/java/#_connection_pooling
**Pooling is not compatible with SUBSCRIBE, UNSUBSCRIBE, PSUBSCRIBE or PUNSUBSCRIBE because these commands will modify the way the connection operates and the connection cannot be reused.**

https://vertx.io/docs/vertx-redis-client/java/#_pubsub_mode

And since all Quarkus producers inside **RedisClientsProducer** class are using the **BaseRedisClient** which has a connection manager of a connection pool, then **ReactiveRedisClient** is not supporting pub/sub.

In my opinion Quarkus should support a setting boolean flag to switch between connection pool mode and single connection mode for vertx redis extension. You can see in RedisAPIImpl class of Vertx:

```
public RedisAPIImpl(RedisConnection connection) {
    this.connection = connection;
    this.redis = null;
  }

  public RedisAPIImpl(Redis redis) {
    this.connection = null;
    this.redis = redis;
  }
  
if (redis != null) {
      // operating in pooled mode
      redis.send(req, promise);
    } else if (connection != null) {
      // operating on connection mode
      connection.send(req, promise);
    }
```
In my opinion Quarkus needs a config flag which of the two constructors to use. Currently Quarkus uses only the 2nd constructor which calls the connection pooling of vertx and vertx doesn't support pub/sub in connection pooling mode.

If Quarkus doesn't plan to fix this, then at least it should be mentioned in the doc.

The workaround currently is to directly use the Vertx redis extension and bypass the Quarkus injectors. I managed to make it work with Mutiny by manually instantiating the Quarkus Redis Mutiny wrappers, but in my opinion it is much simpler to use directly what Vertx provides for pub/sub. Otherwise we need to subscribe to event bus, etc... 

Here is a simple example to use only Vertx:

```
Redis redisClient = Redis.createClient(vertx, "redis://localhost:1234");
    	redisClient
    		.connect()
    	    .onSuccess(conn -> {    
    	    	System.out.println("Successfully connected = " + conn + " " + Thread.currentThread().getName());	
    	    	
    	    	//subscribe(conn);
    	    	
    	    	conn.handler(message -> {
    	    		// do whatever you need to do with your message
    	    		System.out.println("Message = " + message + " " + Thread.currentThread().getName());
	    	    }); 
    	    	
    	    	conn.send(Request.cmd(Command.SUBSCRIBE).arg("CHANNEL1").arg("CHANNEL2"))
    	    	  .onSuccess(res -> {
    	    		  System.out.println("Subscribed");
    	    		        	    		
      	    		  publish(conn);
    	    	  });    	    	    	    	  
    	    })
    	    .onFailure(ex -> {
    	    	System.out.print("Failed to connect " + ex.getMessage());
    		    ex.printStackTrace();
    	    });
    	    
private void publish(RedisConnection conn) {
    	conn.send(Request.cmd(Command.PUBLISH).arg("CHANNEL1").arg("Hello World!"))
	  	  .onSuccess(res -> {
	  		    System.out.println("Published Hello World, res = " + res);
	  		  });
	  	
    	conn.send(Request.cmd(Command.PUBLISH).arg("CHANNEL2").arg("How are you!"))
		  	  .onSuccess(res -> {
		  		    System.out.println("Published How are you!, res = " + res);
		  		  });
  	
    }
```

To conclude - I think Quarkus should take some action - at least mention that in the doc and maybe provide that boolean flag config to be able to configure the injected client **NOT** to use pooling.    

And here is the workaround to use the Mutiny wrappers after manually getting the vertx connection from the callback, so the pooling is bypassed:

```
private void subscribe(RedisConnection connection) {
    	RedisAPI vertxRedisAPI = RedisAPI.api(connection);
    	io.vertx.mutiny.redis.client.RedisAPI redisAPI = new io.vertx.mutiny.redis.client.RedisAPI(vertxRedisAPI);

    	redisAPI.subscribe(List.of("CHANNEL1", "CHANNEL2"))
	    	.subscribe().with(
					response -> { System.out.println("Subscribed"); publish(redisAPI); },
					error -> System.out.println("Error received: " + error));
    }
    
    private void publish(io.vertx.mutiny.redis.client.RedisAPI redisAPI) {
    	redisAPI.publish("CHANNEL1", "message1").subscribe().with(
				response -> System.out.println("Published 1st Message Response: " + response),
				error -> System.out.println("Error received: " + error));
    	
    	redisAPI.publish("CHANNEL2", "message1").subscribe().with(
				response -> System.out.println("Published 2nd Message Response: " + response),
				error -> System.out.println("Error received: " + error));
    }
    
    public void initializeConsumer() {
    	eventBus.<JsonObject>consumer("io.vertx.redis.CHANNEL1", message -> {
    		System.out.println("Consumed event = " + message.body());
    	});
    	
    	eventBus.<JsonObject>consumer("io.vertx.redis.CHANNEL2", message -> {
    		System.out.println("Consumed event = " + message.body());
    	});
    }

```

**Comment by digitalsamba** [comment]
Disclaimer: do NOT publish & consume in same connection like in the above examples - it was for a test. Just open separate connections.

**Comment by tamht298** [comment]
I had same issue, do you solve it?

**Comment by digitalsamba** [comment]
> I had same issue, do you solve it?

You should manage the connection manually, I suppose - Redis.createClient, like in the example I gave above. The Quarkus produced injectors are using connection pooling, which is not compatible with pub/sub.

**Comment by tamht298** [comment]
> > I had same issue, do you solve it?
> 
> You should manage the connection manually, I suppose - Redis.createClient, like in the example I gave above. The Quarkus produced injectors are using connection pooling, which is not compatible with pub/sub.

Oh, thanks so much. Your workaround working properly, but I hope Quarkus team can enhance this feature

**Comment by machi1990** [comment]
> To conclude - I think Quarkus should take some action - at least mention that in the doc and maybe provide that boolean flag config to be able to configure the injected client NOT to use pooling.

Agreed, this should be better documented or made to work transparently without requiring extra work from the consumers of the extension as exposing yet another flag might confuse them. 

> Pooling is not compatible with SUBSCRIBE, UNSUBSCRIBE, PSUBSCRIBE or PUNSUBSCRIBE because these commands will modify the way the connection operates and the connection cannot be reused

@pmlopes I've just became aware of that, is this something new in one of the recent versions of the vertx-redis client? 
I am thinking of opening a new connection that will be used by those commands each time the corresponding method is called, and closing the connection afterwards. Is that the best way to make to go about it? What's your advise here?

**Comment by vundyalaavinash** [comment]
@machi1990 is there an elegant way to handle this? I am also facing the same issue.
Like @digitalsamba has pointed out can we please make it config driven.

**Comment by cescoffier** [comment]
This will be handled in #24750

**Comment by zyc** [comment]
> Today I had same time to dig deeper and found the issue.
> 
> Basically the vertx redis client doesn't support pub/sub in connection pooled mode.
> 
> https://vertx.io/docs/vertx-redis-client/java/#_connection_pooling **Pooling is not compatible with SUBSCRIBE, UNSUBSCRIBE, PSUBSCRIBE or PUNSUBSCRIBE because these commands will modify the way the connection operates and the connection cannot be reused.**
> 
> https://vertx.io/docs/vertx-redis-client/java/#_pubsub_mode
> 
> And since all Quarkus producers inside **RedisClientsProducer** class are using the **BaseRedisClient** which has a connection manager of a connection pool, then **ReactiveRedisClient** is not supporting pub/sub.
> 
> In my opinion Quarkus should support a setting boolean flag to switch between connection pool mode and single connection mode for vertx redis extension. You can see in **RedisAPIImpl** class of Vertx:
> 
> ```
> public RedisAPIImpl(RedisConnection connection) {
>     this.connection = connection;
>     this.redis = null;
>   }
> 
>   public RedisAPIImpl(Redis redis) {
>     this.connection = null;
>     this.redis = redis;
>   }
>   
> if (redis != null) {
>       // operating in pooled mode
>       redis.send(req, promise);
>     } else if (connection != null) {
>       // operating on connection mode
>       connection.send(req, promise);
>     }
> ```
> 
> In my opinion Quarkus needs a config flag which of the two constructors to use. Currently Quarkus uses only the 2nd constructor which calls the connection pooling of vertx and vertx doesn't support pub/sub in connection pooling mode.
> 
> If Quarkus doesn't plan to fix this, then at least it should be mentioned in the doc.
> 
> The workaround currently is to directly use the Vertx redis extension and bypass the Quarkus injectors. I managed to make it work with Mutiny by manually instantiating the Quarkus Redis Mutiny wrappers, but in my opinion it is much simpler to use directly what Vertx provides for pub/sub. Otherwise we need to subscribe to event bus, etc...
> 
> Here is a simple example to use only Vertx:
> 
> ```
> Redis redisClient = Redis.createClient(vertx, "redis://localhost:1234");
>     	redisClient
>     		.connect()
>     	    .onSuccess(conn -> {    
>     	    	System.out.println("Successfully connected = " + conn + " " + Thread.currentThread().getName());	
>     	    	
>     	    	//subscribe(conn);
>     	    	
>     	    	conn.handler(message -> {
>     	    		// do whatever you need to do with your message
>     	    		System.out.println("Message = " + message + " " + Thread.currentThread().getName());
> 	    	    }); 
>     	    	
>     	    	conn.send(Request.cmd(Command.SUBSCRIBE).arg("CHANNEL1").arg("CHANNEL2"))
>     	    	  .onSuccess(res -> {
>     	    		  System.out.println("Subscribed");
>     	    		        	    		
>       	    		  publish(conn);
>     	    	  });    	    	    	    	  
>     	    })
>     	    .onFailure(ex -> {
>     	    	System.out.print("Failed to connect " + ex.getMessage());
>     		    ex.printStackTrace();
>     	    });
>     	    
> private void publish(RedisConnection conn) {
>     	conn.send(Request.cmd(Command.PUBLISH).arg("CHANNEL1").arg("Hello World!"))
> 	  	  .onSuccess(res -> {
> 	  		    System.out.println("Published Hello World, res = " + res);
> 	  		  });
> 	  	
>     	conn.send(Request.cmd(Command.PUBLISH).arg("CHANNEL2").arg("How are you!"))
> 		  	  .onSuccess(res -> {
> 		  		    System.out.println("Published How are you!, res = " + res);
> 		  		  });
>   	
>     }
> ```
> 
> To conclude - I think Quarkus should take some action - at least mention that in the doc and maybe provide that boolean flag config to be able to configure the injected client **NOT** to use pooling.
> 
> And here is the workaround to use the Mutiny wrappers after manually getting the vertx connection from the callback, so the pooling is bypassed:
> 
> ```
> private void subscribe(RedisConnection connection) {
>     	RedisAPI vertxRedisAPI = RedisAPI.api(connection);
>     	io.vertx.mutiny.redis.client.RedisAPI redisAPI = new io.vertx.mutiny.redis.client.RedisAPI(vertxRedisAPI);
> 
>     	redisAPI.subscribe(List.of("CHANNEL1", "CHANNEL2"))
> 	    	.subscribe().with(
> 					response -> { System.out.println("Subscribed"); publish(redisAPI); },
> 					error -> System.out.println("Error received: " + error));
>     }
>     
>     private void publish(io.vertx.mutiny.redis.client.RedisAPI redisAPI) {
>     	redisAPI.publish("CHANNEL1", "message1").subscribe().with(
> 				response -> System.out.println("Published 1st Message Response: " + response),
> 				error -> System.out.println("Error received: " + error));
>     	
>     	redisAPI.publish("CHANNEL2", "message1").subscribe().with(
> 				response -> System.out.println("Published 2nd Message Response: " + response),
> 				error -> System.out.println("Error received: " + error));
>     }
>     
>     public void initializeConsumer() {
>     	eventBus.<JsonObject>consumer("io.vertx.redis.CHANNEL1", message -> {
>     		System.out.println("Consumed event = " + message.body());
>     	});
>     	
>     	eventBus.<JsonObject>consumer("io.vertx.redis.CHANNEL2", message -> {
>     		System.out.println("Consumed event = " + message.body());
>     	});
>     }
> ```

Thanks for that!

**Comment by cescoffier** [comment]
Fixed in #26268
